### DISCUSSION TITLE
multi master HA with Embedded etcd with dns as LB

### DISCUSSION THREAD

**Discussion by migs35323** [k3s-io/k3s#8000] [Q&A] [answered]
I am trying to get fully HA in a new k3s cluster with the embedded etc, 3 masters
i was reading the docs and i found  you can use a dns as a  "load balancer" 
https://docs.k3s.io/datastore/ha#5-optional-configure-a-fixed-registration-address
so i have something like this:
```
mycluster.example.com A 192.168.69.0
mycluster.example.com A 192.168.69.1
mycluster.example.com A 192.168.69.2
```

in the docs https://docs.k3s.io/datastore/ha-embedded the install script mentions we can use a hostname
`curl -sfL https://get.k3s.io | K3S_TOKEN=SECRET sh -s - server --server https://<ip or hostname of server1>:6443`

so, my question is, can i create a cluster with all the nodes server flag pointing to mycluster.example.com dns record? and the masters will be able to sync and talk with each other fine or will it have problems due to a master trying to talk with itself?

the documentation mentions "server1" however i am trying to avoid being dependant on "server1" and having a true multi master system

**Comment by brandond** [comment, marked as answer]
I would probably not recommend using a DNS load-balancer with static address records. It should work eventually, due to RKE2 retrying and receiving a different address from the list each time, but it may take longer than expected for things to come up.

If you're going to do this, I would suggest using a DNS service that supports health checks, as for example Route 53 does with their Alias record type. This functionality is also available from other DNS appliances or server packages.

**Comment by migs35323** [comment]
> I would probably not recommend using a DNS load-balancer with static address records. It should work eventually, due to RKE2 retrying and receiving a different address from the list each time, but it may take longer than expected for things to come up.
> 
> If you're going to do this, I would suggest using a DNS service that supports health checks, as for example Route 53 does with their Alias record type. This functionality is also available from other DNS appliances or server packages.

thank you for your answer

so, if i use the server1 address to join other nodes, when this node1 goes down, will that affect the cluster?
i am wondering if the server1 gets deleted what happens to the cluster since the other masters will have in their configuration the endpoint of server1.
my specific example:
server1 machine gets wipped (k3s uninstall script for example), some network issue happens and server2 goes down, server2 gets up again automatically and tries to boot with the server flag still pointing to server1
i believe in this example, server2 wont be back up again unless we change it's endpoint?

**Reply by brandond** [reply]
> i am wondering if the server1 gets deleted what happens to the cluster since the other masters will have in their configuration the endpoint of server1.

They should be able to start up without issue, even if the original server is offline or otherwise unavailable.

**Comment by migs35323** [comment]
hey @brandond , i am having actually this issue.
the first server node the one that bootstrapped the cluster is gone and i no longer have it.
the 2 other server nodes had the server flag pointed to the first node to start the cluster
so my cluster is being keeped alive by the 2 masters that joined the cluster and it seems to work
however i am trying to recreate the first node and trying to join it into the cluster, i can see it as a new "learned" in the etcd member list however after some time it fails and is no longer in the list, the k3s systemd logs i can see this:

```
Jul 28 09:53:18 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:18.888+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003432c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: authentication handshake failed: read tcp 127.0.0.1:44804->127.0.0.1:2379: i/o timeout\""}
Jul 28 09:53:18 cluster-master-0 k3s[217902]: {"level":"info","ts":"2023-07-28T09:53:18.889+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/client.go:210","msg":"Auto sync endpoints failed.","error":"context deadline exceeded"}
Jul 28 09:53:18 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:18.890+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003432c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: authentication handshake failed: context deadline exceeded\""}
Jul 28 09:53:18 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:18+01:00" level=info msg="Failed to test data store connection: context deadline exceeded"
Jul 28 09:53:19 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:19+01:00" level=info msg="Waiting for API server to become available"
Jul 28 09:53:19 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:19+01:00" level=info msg="Waiting for etcd server to become available"
Jul 28 09:53:21 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:21.667+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003432c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
Jul 28 09:53:21 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:21+01:00" level=error msg="Failed to check local etcd status for learner management: context deadline exceeded"
Jul 28 09:53:21 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:21+01:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"
Jul 28 09:53:22 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:22.478+0100","caller":"etcdserver/server.go:2065","msg":"failed to publish local member to cluster through raft","local-member-id":"nice-id","local-member-attributes":"{Name:cluster-master-0-5fed61dd ClientURLs:[https://192.168.1.20:2379]}","request-path":"/0/members/nice-id/attributes","publish-timeout":"15s","error":"etcdserver: request timed out"}
Jul 28 09:53:26 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:26+01:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"
Jul 28 09:53:31 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:31+01:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"
Jul 28 09:53:33 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:33.890+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003432c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = latest balancer error: last connection error: connection error: desc = \"transport: authentication handshake failed: read tcp 127.0.0.1:44804->127.0.0.1:2379: i/o timeout\""}
Jul 28 09:53:33 cluster-master-0 k3s[217902]: {"level":"info","ts":"2023-07-28T09:53:33.890+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/client.go:210","msg":"Auto sync endpoints failed.","error":"context deadline exceeded"}
Jul 28 09:53:36 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:36.667+0100","logger":"etcd-client","caller":"v3@v3.5.3-k3s1/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003432c0/127.0.0.1:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
Jul 28 09:53:36 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:36+01:00" level=error msg="Failed to check local etcd status for learner management: context deadline exceeded"
Jul 28 09:53:36 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:36+01:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"
Jul 28 09:53:37 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:37.478+0100","caller":"etcdserver/server.go:2065","msg":"failed to publish local member to cluster through raft","local-member-id":"nice-id","local-member-attributes":"{Name:cluster-master-0-5fed61dd ClientURLs:[https://192.168.1.20:2379]}","request-path":"/0/members/nice-id/attributes","publish-timeout":"15s","error":"etcdserver: request timed out"}
Jul 28 09:53:41 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:41+01:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"
Jul 28 09:53:46 cluster-master-0 k3s[217902]: time="2023-07-28T09:53:46+01:00" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: https://127.0.0.1:6443/v1-k3s/readyz: 500 Internal Server Error"
```


this is the script that i am using and that i used in all my server nodes (except the others joined the first node when it existed):
```
curl -sfL https://get.k3s.io | K3S_TOKEN={REDACTED}::server:{REDACTED} INSTALL_K3S_VERSION=v1.25.11+k3s1 INSTALL_K3S_EXEC="server --server https://192.168.69.2:6443 --flannel-iface nebula1 --node-external-ip {THIS_NODE_IP} --node-ip {THIS_NODE_IP} --disable traefik --prefer-bundled-bin --tls-san mycluster.example.com" sh -
```

**Reply by brandond** [reply]
`Jul 28 09:53:37 cluster-master-0 k3s[217902]: {"level":"warn","ts":"2023-07-28T09:53:37.478+0100","caller":"etcdserver/server.go:2065","msg":"failed to publish local member to cluster through raft","local-member-id":"nice-id","local-member-attributes":"{Name:cluster-master-0-5fed61dd ClientURLs:[https://192.168.1.20:2379]}","request-path":"/0/members/nice-id/attributes","publish-timeout":"15s","error":"etcdserver: request timed out"}
`

Are you sure that all the correct ports are open for etcd? It's complaining that it can't communicate with other cluster members.

**Reply by migs35323** [reply]
thank you for having a look, well, i didnt had changes in the network and i believe the ports where open since as i said before this machine was used to bootstrap the cluster and the other nodes joined to it just fine
anyway, after a weekend break, i tried it again and it seems to work now, maybe some network issue was happening or maybe the old node was still to be garbage collected or something... since i didn't changed its name or anything else.. i did manually remove it from the etcd when i did the k3s uninstall when i was having issues starting it... maybe something happened there...
